### DISCUSSION TITLE
[4일차-지원] HTML이 렌더링 중에 JS가 실행되면 멈추는데 그 이유는 무엇인가?

### DISCUSSION THREAD

**Discussion by smarfy99** [zzangCS/zzangInterview#18] [프론트 면접 질문]
제목과 동일

**Comment by ez1n** [comment]
렌더링 과정 중에 <script> 태그를 만나게 되면 진행중인 HTML 파싱을 중지하고 js엔진으로 제어 권한을 넘기기 때문
 js파싱 및 실행이 종료되면 렌더링 엔진으로 다시 돌아가 파싱이 중단된 시점부터 다시 파싱을 시작한다.

**Reply by ez1n** [reply]
** 브라우저 동작 원리와 연결되어 있음! **

> 자바스크립트는 자바스크립트 엔진이 처리하기 때문에 HTML 파서는 script 태그를 만나면 JS 코드 실행을 위해 DOM 생성 프로세스를 중지하고 자바스크립트 엔진으로 권한을 넘긴다.

1. script 태그를 만나면 JS 코드 실행을 위해 DOM 생성 프로세스를 중지하고 HTML 파서에서 자바스크립트 엔진으로 권한이 넘어간다.
2. 자바스크립트 엔진은 script 태그 내의 JS 코드나 src 속성에 정의된 JS 파일을 로드해 파싱 후 실행한다.
3. 실행이 완료되면 다시 HTML 파서로 제어 권한을 넘겨 중지했던 시점으로 돌아가 DOM 생성을 재개한다!

**Comment by smarfy99** [comment]
### HTML 렌더링 중 JS를 만나게되면, 

1. HTML 파서는 스크립트 태그를 만나면 렌더링 제어 권한을 JS 엔진에 넘겨주게 되고, 문서 파싱을 중단한다.
2. JS 엔진이 파싱을 끝내면 다시 html 렌더링 과정으로 돌아가 중단된 시점부터 파싱을 다시 시작한다.
-> 이렇기 때문에 스크립트 태그는 바디 끝에 위치하는 걸 권장한다.

바디 끝에 위치할 경우 HTML문서를 화면에 표시하는 속도가 빨라지고, 사용자가 뷰를 보는데 필요한 웬만한 문서를 해석한 상태기 때문에 사용자 불편이 감소한다.

### 스크립트 파일을 먼저 로드한다면?

스크립트가 파싱되는 동안 html, css 같은 브라우저 문서 작업 수행이 어려워 제대로된 화면의 레이아웃이 구성되지 않은 채로 사용자에게 보여지게 된다.
